src/cli/telemetry.ts

```diff
@@ -0,0 +1,155 @@
+    pageview(page: string): void;
+    timing(category: string, variable: string, time: number): void;
+    event(category: string, action: string, label?: string, value?: string | number): void;
+}
+
+class NoAnalytics implements Analytics {
+    pageview(_page: string) {
+        // Pass
+    }
+
+    timing(_category: string, _variable: string, _time: number) {
+        // Pass
+    }
+
+    event(_category: string, _action: string, _label?: string, _value?: string | number) {
+        // Pass
+    }
+}
+
+class GoogleAnalytics implements Analytics {
+    private readonly visitor: ga.Visitor;
+
+    constructor() {
+        this.visitor = ga(GoogleAnalyticsID, Storage.userId);
+    }
+
+    pageview(page: string): void {
+        this.visitor.pageview(page).send();
+    }
+
+    timing(category: string, variable: string, time: number) {
+        this.visitor.timing(category, variable, time);
+    }
+
+    event(category: string, action: string, label?: string, value?: string | number) {
+        if (label !== undefined) {
+            if (value !== undefined) {
+                this.visitor.event(category, action, label, value).send();
+            } else {
+                this.visitor.event(category, action, label).send();
+            }
+        } else {
+            this.visitor.event(category, action).send();
+        }
+    }
+}
+
+export class Telemetry implements Analytics {
+    private analytics = new NoAnalytics();
+
+    async init() {
+        await storage.init();
+        if (Storage.analyticsState === "enabled") {
+            this.enable();
+        }
+    }
+
+    pageview(page: string): void {
+        this.analytics.pageview(page);
+    }
+
+    timing(category: string, variable: string, time: number) {
+        this.analytics.timing(category, variable, time);
+    }
+
+    event(category: string, action: string, label?: string, value?: string | number) {
+        this.analytics.event(category, action, label, value);
+    }
+
+    enable() {
+        Storage.analyticsState = "enabled";
+        this.analytics = new GoogleAnalytics();
+    }
+
+    disable() {
+        Storage.analyticsState = "disabled";
+        this.analytics = new NoAnalytics();
+    }
+
+    get state(): TelemetryState {
+        return Storage.analyticsState;
+    }
+
+    set state(state: TelemetryState) {
+        Storage.analyticsState = state;
+    }
+
+    async timeAsync<T>(variable: string, work: () => Promise<T>): Promise<T> {
+        const start = new Date().getTime();
+        const result = await work();
+        const end = new Date().getTime();
+        this.timing("default", variable, end - start);
+        return result;
+    }
+}
+
+class Storage {
+    static get<T>(name: string, def: T): T {
+        return storage.getItemSync(name) || def;
+    }
+
+    static set<T>(name: string, val: T) {
+        storage.setItemSync(name, val);
+    }
+
+    static get userId(): string {
+        return Storage.get("userId", uuid());
```

Comment:
This seems weird.  `uuid()` generates a new UUID each time, right?  So if the user ID isn't set, this will return a different one each time?

Also I'd like to use all-caps for acronyms.

Reply:
Oops, I forgot that I removed the code that sets the default value in `get`, well restore.

Reply:
`get` must not have side effects. The debugger will feel free to call it anytime, and it will screw you up.

Reply:
The method is just called `get`, it's not a property.